pkg/services/rpcsrv/server.go

```diff
@@ -3202,3 +3204,102 @@ func (s *Server) getRawNotaryTransaction(reqParams params.Params) (any, *neorpc.
 	}
 	return tx.Bytes(), nil
 }
+
+// getBlockNotifications returns notifications from a specific block with optional filtering.
+func (s *Server) getBlockNotifications(reqParams params.Params) (any, *neorpc.Error) {
+	param := reqParams.Value(0)
+	hash, respErr := s.blockHashFromParam(param)
+	if respErr != nil {
+		return nil, respErr
+	}
+
+	block, err := s.chain.GetBlock(hash)
+	if err != nil {
+		return nil, neorpc.ErrUnknownBlock
+	}
+
+	var filter *neorpc.NotificationFilter
+	if len(reqParams) > 1 {
+		filter = new(neorpc.NotificationFilter)
+		err := json.Unmarshal(reqParams[1].RawMessage, filter)
+		if err != nil {
+			return nil, neorpc.WrapErrorWithData(neorpc.ErrInvalidParams, fmt.Sprintf("invalid filter: %s", err))
+		}
+		if err := filter.IsValid(); err != nil {
+			return nil, neorpc.WrapErrorWithData(neorpc.ErrInvalidParams, fmt.Sprintf("invalid filter: %s", err))
+		}
+	}
+
+	var notifications []state.ContainedNotificationEvent
+	for _, tx := range block.Transactions {
+		aers, err := s.chain.GetAppExecResults(tx.Hash(), trigger.Application)
+		if err != nil {
+			continue
+		}
+		for _, aer := range aers {
+			if aer.VMState == vmstate.Halt {
+				for _, evt := range aer.Events {
+					ntf := state.ContainedNotificationEvent{
+						Container:         aer.Container,
+						NotificationEvent: evt,
+					}
+					if filter == nil || rpcevent.Matches(&testComparator{
+						id:     neorpc.NotificationEventID,
+						filter: *filter,
+					}, &testContainer{ntf: &ntf}) {
+						notifications = append(notifications, ntf)
+					}
+				}
+			}
+		}
+	}
+
+	// Also check for notifications from the block itself (PostPersist)
```

Comment:
You need to keep the sequence correct and check for `trigger.OnPersist` _before_ digging into transaction list.

Reply:
We may additionally extend `neorpc.NotificaitonFilter` with trigger parameter, it may be useful in some cases, what do you think?

Reply:
What are the use cases? People interested in block-level things (and not a lot happens there) probably pull block applogs directly and otherwise it's all `Application`.